lib/runtime/storage/trie.go

```diff
@@ -17,97 +18,111 @@ import (
 // TrieState is a wrapper around a transient trie that is used during the course of executing some runtime call.
 // If the execution of the call is successful, the trie will be saved in the StorageState.
 type TrieState struct {
-	t       *trie.Trie
-	oldTrie *trie.Trie // this is the trie before BeginStorageTransaction is called. set to nil if it isn't called
-	lock    sync.RWMutex
+	mtx          sync.RWMutex
+	transactions *list.List
 }
 
-// NewTrieState returns a new TrieState with the given trie
-func NewTrieState(t *trie.Trie) *TrieState {
-	if t == nil {
-		t = trie.NewEmptyTrie()
-	}
-
+func NewTrieState(state *trie.Trie) *TrieState {
+	transactions := list.New()
+	transactions.PushBack(state)
 	return &TrieState{
-		t: t,
+		transactions: transactions,
 	}
 }
 
-// Trie returns the TrieState's underlying trie
-func (s *TrieState) Trie() *trie.Trie {
-	return s.t
+func (t *TrieState) getCurrentTrie() *trie.Trie {
+	return t.transactions.Back().Value.(*trie.Trie)
 }
 
-// Snapshot creates a new "version" of the trie. The trie before Snapshot is called
-// can no longer be modified, all further changes are on a new "version" of the trie.
-// It returns the new version of the trie.
-func (s *TrieState) Snapshot() *trie.Trie {
-	return s.t.Snapshot()
+func (t *TrieState) updateCurrentTrie(new *trie.Trie) {
+	t.transactions.Back().Value = new
 }
 
-// BeginStorageTransaction begins a new nested storage transaction
+// StartTransaction begins a new nested storage transaction
 // which will either be committed or rolled back at a later time.
-func (s *TrieState) BeginStorageTransaction() {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	s.oldTrie = s.t
-	s.t = s.t.Snapshot()
+func (t *TrieState) StartTransaction() {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	t.transactions.PushBack(t.getCurrentTrie().Snapshot())
+}
+
+// Rollback rolls back all storage changes made since StartTransaction was called.
+func (t *TrieState) Rollback() {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	if t.transactions.Len() <= 1 {
+		panic("no transactions to rollback")
```

Comment:
Is this a situation that we need to panic in?

Reply:
Yeah, committing or rollback without an ongoing transaction will lead to a corrupt state